epsagon/aws_sdk_v2_factories/sns_factory.go

```diff
@@ -0,0 +1,46 @@
+package epsagonawsv2factories
+
+import (
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/epsagon/epsagon-go/protocol"
+	"reflect"
+	"strings"
+)
+
+// SnsEventDataFactory creates an Epsagon Resource from aws.Request
+func SnsEventDataFactory(r *aws.Request, res *protocol.Resource, metadataOnly bool) {
+	if !metadataOnly {
+		inputValue := reflect.ValueOf(r.Params).Elem()
+		updateMetadataFromValue(inputValue, "Message", "Notification Message", res.Metadata)
+	}
+	handleSpecificOperation(r, res, metadataOnly,
+		map[string]specificOperationHandler{
+			"CreateTopic": handleSNSCreateTopic,
+			"Publish":     handlerSNSPublish,
+		},
+		handleSNSdefault,
+	)
+}
+
+func handleSNSdefault(r *aws.Request, res *protocol.Resource, metadataOnly bool) {
+	inputValue := reflect.ValueOf(r.Params).Elem()
+	topicArn, ok := getFieldStringPtr(inputValue, "TopicArn")
+	if ok {
+		splitTopic := strings.Split(topicArn, ":")
```

Comment:
This get ARN resource name is repeating itself and can be extracted to a util function. That would also handle the index out of bounds in SFN 🙂